03-JavaServiceSkeleton/MessageConsumerApplication/app/src/main/java/com/example/shevchenko/message/MainActivity.java

```diff
@@ -0,0 +1,62 @@
+package com.example.shevchenko.message;
+
+import android.content.ComponentName;
+import android.content.Intent;
+import android.content.ServiceConnection;
+import android.os.IBinder;
+import android.os.RemoteException;
+import android.support.v7.app.AppCompatActivity;
+import android.os.Bundle;
+import android.util.Log;
+import android.view.View;
+import android.widget.Button;
+import android.widget.EditText;
+
+import com.example.shevchenko.message.service.IMessageAidlInterface;
+
+public class MainActivity extends AppCompatActivity {
+    private final static String TAG = "MessageConsumerApplication";
+    private IMessageAidlInterface messageService;
+
+    private ServiceConnection serviceConnection = new ServiceConnection() {
+        @Override
+        public void onServiceConnected(ComponentName name, IBinder service) {
+            if (service != null) {
+                messageService = IMessageAidlInterface.Stub.asInterface(service);
+            }
+        }
+
+        @Override
+        public void onServiceDisconnected(ComponentName name) {
+
+        }
+    };
+
+    @Override
+    protected void onCreate(Bundle savedInstanceState) {
+        super.onCreate(savedInstanceState);
+        setContentView(R.layout.activity_main);
+
+        final EditText edtMessage = (EditText) findViewById(R.id.edtMessage);
+        final Button btnReceive = (Button) findViewById(R.id.btnReceive);
+        btnReceive.setOnClickListener(new View.OnClickListener() {
+            @Override
+            public void onClick(View v) {
+                try {
+                    String message = messageService.consumeMessage();
+                    edtMessage.setText(message);
+                }catch (RemoteException | NullPointerException e) {
+                    Log.e(TAG, "Receive is failed");
+                }
+            }
+        });
+
+        Intent intent = new Intent();
+        intent.setClassName("com.example.shevchenko.message.service", "com.example.shevchenko.message.service.MessageService");
+        try {
+            bindService(intent, serviceConnection, BIND_AUTO_CREATE);
```

Comment:
Where is unbindService() ?

Reply:
Fixed in the commit 6a82d4311abfa610aa43a3dadde40fd9a47a052f